### DISCUSSION TITLE
Evennia 1.0 - PyCharm attach to process debugging failure

### DISCUSSION THREAD

**Discussion by MardMallard** [evennia/evennia#2878] [General Discussion]
### **This is a resolved issue, but I would like to document the resolution here to help others.**

I recently upgraded my Evennia project to 1.0 following the instructions posted [here](https://github.com/evennia/evennia/discussions/2640) and afterwards was unable to use PyCharm's _attach to process_ feature to debug my code. 

**The problem:**

[Attach to process](https://www.jetbrains.com/help/pycharm/attaching-to-local-process.html) is the recommended way to debug in Evennia as per the [Setting up PyCharm with Evennia](https://www.evennia.com/docs/1.0-dev/Coding/Setting-up-PyCharm.html) guide. When using attach to process you must choose the twistd process with the server.py parameter, which in my experience there are usually two. Usually, one of these should work.

![IIYEWEX](https://user-images.githubusercontent.com/5182101/193392790-ff485179-23aa-410e-9e65-5c756b38d0ea.png)

After upgrading, neither options worked at all. Attempting to attach to one process would result in a timeout. The other would provide a cryptic error message:

```Traceback (most recent call last):
  File "C:\Program Files\JetBrains\PyCharm Community Edition 2021.3.3\plugins\python-ce\helpers\pydev\pydevd_attach_to_process\attach_pydevd.py", line 88, in <module>
    main(process_command_line(sys.argv[1:]))
  File "C:\Program Files\JetBrains\PyCharm Community Edition 2021.3.3\plugins\python-ce\helpers\pydev\pydevd_attach_to_process\attach_pydevd.py", line 83, in main
    add_code_to_python_process.run_python_code(
  File "C:\Program Files\JetBrains\PyCharm Community Edition 2021.3.3\plugins\python-ce\helpers\pydev\pydevd_attach_to_process\add_code_to_python_process.py", line 306, in run_python_code_windows
    assert resolve_label(process, b'PyGILState_Ensure')
  File "C:\Program Files\JetBrains\PyCharm Community Edition 2021.3.3\plugins\python-ce\helpers\pydev\pydevd_attach_to_process\add_code_to_python_process.py", line 266, in resolve_label
    address = process.resolve_label(label)
  File "C:\Program Files\JetBrains\PyCharm Community Edition 2021.3.3\plugins\python-ce\helpers\pydev\pydevd_attach_to_process\winappdbg\module.py", line 1546, in resolve_label
    address = self.resolve_label_components(module, function, offset)
  File "C:\Program Files\JetBrains\PyCharm Community Edition 2021.3.3\plugins\python-ce\helpers\pydev\pydevd_attach_to_process\winappdbg\module.py", line 1635, in resolve_label_components
    raise RuntimeError(msg)
RuntimeError: Function b'PyGILState_Ensure' not found in any module
```

This error is a complete red herring and likely to lead you into Google Hell. It's actually normal for this one to not work. The one that is supposed to work is the one that times out.

**The fix:**

For reasons that are arcane to me, I resolved this issue by downgrading my Python to version 3.9.13. Python versions below 3.9 all work correctly with _Attach to Process_, however Evennia won't start unless you have version 3.9 or above. Be aware that there may have unintended side-effects. The instructions for installing Evennia 1.0 specifically request Python 3.10 and I assume that there is a good reason for that. I'm making this trade-off because I find it difficult to work without a debugger. 

If downgrading your Python version doesn't fix your issue, make sure that you have tried using Attach to Process on _all_ twistd processes with a server.py parameter (there can be multiple). Otherwise, here is my exact installation:

Python: 3.9.13
PyCharm: 2021.3.3 Build #PC-213.7172.26, built on March 16, 2022
OS: Microsoft Windows 10 Home v10.0.19044 Build 19044

Hope this helps anyone else running into this issue. Any insights etc are welcome. I won't create a GitHub issue for this since it is not an issue with Evennia's code.

**Comment by BigJMoney** [comment]
Just to update for a modern audience :)

I ran into this as well on Windows. It appears to have been an issue with older versions of Pycharm and Python 3.10+, but it was fixed:
https://youtrack.jetbrains.com/issue/PY-48037/Support-attach-to-process-on-Windows-with-Python-3.10

On the latest version of PyCharm (2025.3), I'm able to attach the debugger using Python 3.11. I still need to test how well the debugging works, though. Also note that someone left a comment in that bug report that says the issue is occurring again with Python 3.12. I have not confirmed that, myself. Maybe a good reason to switch away from PyCharm for Evennia :/. I haven't tried VS Code, so I don't know if that debugger works or not.

**Comment by electroglyph** [comment]
for vscode this is my preferred way to debug: https://github.com/evennia/evennia/tree/main/evennia/contrib/utils/debugpy

you get the full GUI debugger experience that way with breakpoints, ability to look at variable contents, etc.
